KEP-4974 endpoints deprecation blog post for 1.33 (#49976)
kubernetes/website: content/en/blog/_posts/2025-XX-XX-endpoints-deprecation.md

---
layout: blog

> title: "Endpoints formally deprecated in favor of EndpointSlices"

Reviewer: With this title, we can only publish it on the day that we formally announce / confirm deprecation.

Maybe we want to have two articles:
1. how to adopt EndpointSlice and stop using Endpoints
1. Endpoints is now formally deprecated, go read the other article to find out what to do

?

Author: changed the title to be less authoritative-sounding

slug: endpoints-deprecation
date: 2025-XX-XX
draft: true
author: >
  Dan Winship (Red Hat)
---

> Since the addition of [EndpointSlices] several years ago, the

Reviewer 2: I think it would be useful to give a bit more detail as to how long it has been since the addition.

```suggestion
Since the addition of [EndpointSlices] as alpha in v1.15 and later GA in v1.21, the
```

We could also potentially link to the original KEP https://github.com/kubernetes/enhancements/issues/752

Endpoints API in Kubernetes has been gathering dust. New Service
features like [dual-stack networking] and [topology-aware routing] are
only supported via the EndpointSlice API, so all service proxies,
Gateway API implementations, and similar controllers have had to be
ported from using Endpoints to using EndpointSlices. At this point,
the Endpoints API is really only there to avoid breaking end user
workloads and scripts that still make use of it.

> We are now beginning the process of officially deprecating the

Reviewer: Avoid “We“ when there is a single author. See https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we and also the provisos in https://kubernetes.io/docs/contribute/review/reviewing-prs/#blog

> We are now beginning the process of officially deprecating the
> Endpoints API, with clarifications to the documentation, and warnings
> from the API server when a client reads or writes Endpoints resources.

Reviewer 2: ```suggestion
We are deprecating the Endpoints API. This will initially involve documentation updates and API server warnings when Endpoints resources are accessed.
```

Wording simplified slightly

Reviewer: Also see https://github.com/kubernetes/website/pull/49976/files#r2007454583

Author: That rewording seems a little abrupt. I tried rearranging and rewording some things...

> While the Endpoints type itself can never go away as long as the "v1"

Reviewer: ```suggestion
While the Endpoints type itself can never go away as long as the `v1` Endpoints
```

We technically can drop Endpoints v1 without dropping all v1 APIs (would need a major Kubernetes version, but if we ever do a major version bump, our compatibility guarantees allow doing anything we like with the APIs).

Author: That basically says "Endpoints can't go away as long as Endpoints hasn't gone away".

I couldn't come up with an entirely correct explanation that didn't get mired in irrelevant details, so I just changed it to say Endpoints will "probably" never go away.

API still exists in Kubernetes, we intend to eventually change the
[Kubernetes Conformance] criteria to no longer require that clusters
run the _Endpoints controller_ (which generates Endpoints objects
based on Services and Pods) and the _EndpointSlice mirroring
controller_ (which copies user-created Endpoints objects to equivalent
EndpointSlice objects), to avoid doing work that is unneeded in most
modern-day clusters.

> Thus, although there are no changes to functionality or supportedness
> in Kubernetes 1.33, we recommend that all users who still have
> workloads or scripts that use the Endpoints API should start thinking
> about migrating them to EndpointSlices.

Reviewer 2: ```suggestion
Thus, although there are no changes to functionality or supportedness
in Kubernetes 1.33, we recommend that all users who still have
workloads or scripts that use the Endpoints API should start migrating
them to EndpointSlices.
```

[EndpointSlices]: /blog/2020/09/02/scaling-kubernetes-networking-with-endpointslices/
[dual-stack networking]: /docs/concepts/services-networking/dual-stack/
[topology-aware routing]: /docs/concepts/services-networking/topology-aware-routing/
[Kubernetes Conformance]: https://www.cncf.io/training/certification/software-conformance/

## Notes on migrating from Endpoints to EndpointSlices

For end users, the biggest change between the Endpoints API and the
EndpointSlice API is that while every Service with a `selector` has
exactly 1 Endpoints object (with the same name as the Service), a
Service may have any number of EndpointSlices associated with it:

> ```bash

Reviewer: If including the prompt, change `bash` to `console`

$ kubectl get endpoints myservice
NAME        ENDPOINTS          AGE
myservice   10.180.3.17:443    1h

$ kubectl get endpointslice -l kubernetes.io/service-name=myservice
NAME              ADDRESSTYPE   PORTS   ENDPOINTS          AGE
myservice-7vzhx   IPv4          443     10.180.3.17        21s
myservice-jcv8s   IPv6          443     2001:db8:0123::5   21s
```

In this case, because the service is dual stack, it has 2
EndpointSlices: 1 for IPv4 addresses and 1 for IPv6 addresses. (The
Endpoints API does not support dual stack, so the Endpoints object
shows only the addresses in the cluster's primary address family.) A
Service can also have multiple EndpointSlices in other cases; in
particular, if it has more than 100 endpoint pods, then the endpoints
will be split up into multiple EndpointSlice objects, which can be
processed more efficiently than a single excessively-large Endpoints
object.

Because there is not a predictable 1-to-1 mapping between Services and
EndpointSlices, there is no way to know what the actual name of the
EndpointSlice resource(s) will be ahead of time; thus, instead of
fetching the EndpointSlice(s) by name, you instead ask for all

> EndpointSlices with a "`kubernetes.io/service-name`" label pointing to

Reviewer: ```suggestion
EndpointSlices with a `kubernetes.io/service-name` [label](/docs/concepts/overview/working-with-objects/labels/) pointing to
```

For end users, the biggest change between the Endpoints API and the
EndpointSlice API is that while every Service with a `selector` has
exactly 1 Endpoints object (with the same name as the Service), a
Service may have any number of EndpointSlices associated with it:

```console
$ kubectl get endpoints myservice
NAME        ENDPOINTS          AGE
myservice   10.180.3.17:443    1h

$ kubectl get endpointslice -l kubernetes.io/service-name=myservice
NAME              ADDRESSTYPE   PORTS   ENDPOINTS          AGE
myservice-7vzhx   IPv4          443     10.180.3.17        21s
myservice-jcv8s   IPv6          443     2001:db8:0123::5   21s
```

In this case, because the service is dual stack, it has 2
EndpointSlices: 1 for IPv4 addresses and 1 for IPv6 addresses. (The
Endpoints API does not support dual stack, so the Endpoints object
shows only the addresses in the cluster's primary address family.) A
Service can also have multiple EndpointSlices in other cases; in
particular, if it has more than 100 endpoint pods, then the endpoints
will be split up into multiple EndpointSlice objects, which can be
processed more efficiently than a single excessively-large Endpoints
object.

> Because there is not a predictable 1-to-1 mapping between Services and

Reviewer 3: Would it be possible to expand on this section? There are a bunch of posts on discuss about a use case for creating endpoint manifests in regard to services without a selector:
- https://discuss.kubernetes.io/t/understanding-service-without-selector-and-endpointslice-behavior-in-kubernetes
- https://discuss.kubernetes.io/t/servicetopology-and-manually-endpointslice
- https://discuss.kubernetes.io/t/use-https-protocol-for-endpoints-in-kubernetes-services
- https://discuss.kubernetes.io/t/pod-access-outside-cluster

To note, I do see the [docs](https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors) already provide the approach with slices, but just wanted to at least highlight the use case. What I am hoping for, similar to the go example, is a manifest of Endpoints and the equivalent of an Endpoint slices.

Reviewer 4: @danquack would you be willing to write that article? I think the main obstacle to publication is finding a volunteer author.

Reviewer 3: @lmktfy I was actually thinking it could be apart of this blog post. Happy to provide a snippet if @danwinship thinks that would be helpful?

---
layout: blog
title: "Continuing the transition from Endpoints to EndpointSlices"
slug: endpoints-deprecation
date: 2025-XX-XX
draft: true
author: >
  Dan Winship (Red Hat)
---

Since the addition of [EndpointSlices] ([KEP-752]) as alpha in v1.15
and later GA in v1.21, the
Endpoints API in Kubernetes has been gathering dust. New Service
features like [dual-stack networking] and [traffic distribution] are
only supported via the EndpointSlice API, so all service proxies,
Gateway API implementations, and similar controllers have had to be
ported from using Endpoints to using EndpointSlices. At this point,
the Endpoints API is really only there to avoid breaking end user
workloads and scripts that still make use of it.

As of Kubernetes 1.33, the Endpoints API is now officially deprecated,
and the API server will return warnings to users who read or write
Endpoints resources rather than using EndpointSlices.

Eventually, the plan (as documented in [KEP-4794]) is to change the
[Kubernetes Conformance] criteria to no longer require that clusters
run the _Endpoints controller_ (which generates Endpoints objects

> based on Services and Pods) and the _EndpointSlice mirroring

Reviewer 5: for the Endpointss controller the decision was clear and unanimous but not for the mirroring controller https://github.com/kubernetes/enhancements/pull/4975#discussion_r1922458035.

I think we should try to fail safe based on past experience with deprecations, is very unlikely the entire ecosystem changes and not having the mirroring controller will break clusters. 

Do you mind omitting the refrence to `the _EndpointSlice mirroring` on purpose, it is also an internal detail most people will not know it exists

Author: Sure (though ftr the discussion in the KEP was mostly about whether we could disable those controllers _by default_ rather than whether we could change conformance to give admins the _option_ to disable them).

Reviewer 5: Conformance kind of means things enabled by default to guarantee portability... If y develop some project that create custom endpoints it will work in all conformance cluster because there has to be something that mirror the to slices in order to work ...

Reviewer 5: I didn't understand that sorry, and I realized my mistake 
... I agree we moved the e2e and conformance tests to not depends on endpoints, and that is the right thing

only supported via the EndpointSlice API, so all service proxies,
Gateway API implementations, and similar controllers have had to be
ported from using Endpoints to using EndpointSlices. At this point,
the Endpoints API is really only there to avoid breaking end user
workloads and scripts that still make use of it.

While the [Kubernetes deprecation policy] means that the Endpoints
type itself will probably never completely go away, the 1.33 release
will take a few more steps towards fully deprecating it, with
clarifications to the documentation, and warnings from the API server
when a client reads or writes Endpoints resources rather than using
EndpointSlices.

Eventually, the plan (as documented in [KEP-4794]) is to change the
[Kubernetes Conformance] criteria to no longer require that clusters
run the _Endpoints controller_ (which generates Endpoints objects
based on Services and Pods) and the _EndpointSlice mirroring
controller_ (which copies user-created Endpoints objects to equivalent
EndpointSlice objects), to avoid doing work that is unneeded in most
modern-day clusters.

Thus, although there are no changes to functionality or supportedness

> in Kubernetes 1.33, users who still have workloads or scripts that use

Reviewer 4: None of these sentences says "The Endpoints API is deprecated". I think we should state that, even though the formal deprecation is part of the release announcement.
